File: src/main/kotlin/com/cfm/search/SearchService.kt
package com.cfm.search

import com.cfm.metadata.MetadataService
import com.cfm.model.FileModel
import kotlinx.serialization.json.Json
import java.util.*

/**
 * Service for searching files based on metadata and content
 */
class SearchService(private val metadataService: MetadataService) {

    /**
     * Searches for files based on a search query and optional filters
     *
     * @param query The search query
     * @param filters Optional filters to apply
     * @return List of matching files
     */
    suspend fun searchFiles(query: String, filters: SearchFilters = SearchFilters()): List<FileModel> {
        // Get all files that match the basic criteria
        val files = metadataService.searchFiles(
            searchTerm = query,
            tags = filters.tags,
            mimeTypes = filters.mimeTypes
        )

        // Apply additional filtering and ranking
        return filterAndRankResults(files, query, filters)
    }

    /**
     * Applies additional filtering and ranks search results
     *
     * @param files The initial list of files
     * @param query The search query
     * @param filters The search filters
     * @return Filtered and ranked list of files
     */
    private fun filterAndRankResults(
        files: List<FileModel>,
        query: String,
        filters: SearchFilters
    ): List<FileModel> {
        // Apply date filters
        var filteredFiles = files.filter { file ->
            val uploadTime = try {
                Date.from(java.time.Instant.parse(file.uploadedAt))
            } catch (e: Exception) {
                Date()
            }

            val afterFilter = filters.uploadedAfter?.let { uploadTime.after(it) } ?: true
            val beforeFilter = filters.uploadedBefore?.let { uploadTime.before(it) } ?: true

            afterFilter && beforeFilter
        }

        // Apply size filters
        filteredFiles = filteredFiles.filter { file ->
            val minSizeFilter = filters.minSize?.let { file.size >= it } ?: true
            val maxSizeFilter = filters.maxSize?.let { file.size <= it } ?: true

            minSizeFilter && maxSizeFilter
        }

        // Apply category filters if specified
        if (filters.categories.isNotEmpty()) {
            filteredFiles = filteredFiles.filter { file ->
                file.metadata.categories.any { category ->
                    filters.categories.any { it.equals(category, ignoreCase = true) }
                }
            }
        }

        // Calculate relevance score for each file and sort by it
        return filteredFiles.map { file ->
            val relevanceScore = calculateRelevanceScore(file, query)
            Pair(file, relevanceScore)
        }
            .sortedByDescending { it.second }
            .map { it.first }
    }

    /**
     * Calculates a relevance score for a file based on how well it matches the query
     *
     * @param file The file model
     * @param query The search query
     * @return A relevance score (higher is better)
     */
    private fun calculateRelevanceScore(file: FileModel, query: String): Double {
        if (query.isEmpty()) {
            return 1.0 // No query means all files are equally relevant
        }

        var score = 0.0
        val lowerQuery = query.lowercase()
        val queryTerms = lowerQuery.split(Regex("\\s+"))

        // Check filename match (highest weight)
        val fileNameLower = file.name.lowercase()
        if (fileNameLower.contains(lowerQuery)) {
            score += 10.0
        }

        // Add points for each query term in the filename
        queryTerms.forEach { term ->
            if (fileNameLower.contains(term)) {
                score += 5.0
            }
        }

        // Check tags match
        file.tags.forEach { tag ->
            if (tag.lowercase().contains(lowerQuery)) {
                score += 3.0
            }

            queryTerms.forEach { term ->
                if (tag.lowercase().contains(term)) {
                    score += 1.5
                }
            }
        }

        // Check categories match
        file.metadata.categories.forEach { category ->
            if (category.lowercase().contains(lowerQuery)) {
                score += 3.0
            }

            queryTerms.forEach { term ->
                if (category.lowercase().contains(term)) {
                    score += 1.5
                }
            }
        }

        // Check extracted text match
        file.metadata.extractedText?.let { text ->
            val lowerText = text.lowercase()

            // Add points for exact matches
            val exactMatches = lowerQuery.toRegex().findAll(lowerText).count()
            score += exactMatches * 2.0

            // Add points for individual term matches
            queryTerms.forEach { term ->
                val termMatches = term.toRegex().findAll(lowerText).count()
                score += termMatches * 0.5
            }
        }

        // Check entity matches
        file.metadata.entities.forEach { entity ->
            if (entity.text.lowercase().contains(lowerQuery)) {
                score += 2.0
            }

            queryTerms.forEach { term ->
                if (entity.text.lowercase().contains(term)) {
                    score += 1.0
                }
            }
        }

        return score
    }
}

/**
 * Filters that can be applied to search results
 */
data class SearchFilters(
    val tags: List<String> = emptyList(),
    val categories: List<String> = emptyList(),
    val mimeTypes: List<String> = emptyList(),
    val uploadedAfter: Date? = null,
    val uploadedBefore: Date? = null,
    val minSize: Long? = null,
    val maxSize: Long? = null
)

File: cloud_file_manager/search/search_service.py
import datetime
import logging
import re
from dataclasses import dataclass
from datetime import datetime
from typing import Dict, List, Optional

from cloud_file_manager.metadata.metadata_service import MetadataService
from cloud_file_manager.models.file_models import FileModel

logger = logging.getLogger(__name__)


@dataclass
class SearchFilters:
    """
    Filters that can be applied to search results
    """
    tags: List[str] = None
    categories: List[str] = None
    mime_types: List[str] = None
    uploaded_after: Optional[datetime] = None
    uploaded_before: Optional[datetime] = None
    min_size: Optional[int] = None
    max_size: Optional[int] = None

    def __post_init__(self):
        # Initialize empty lists
        if self.tags is None:
            self.tags = []
        if self.categories is None:
            self.categories = []
        if self.mime_types is None:
            self.mime_types = []


class SearchService:
    """Service for searching files based on metadata and content"""

    def __init__(self, metadata_service: MetadataService):
        self.metadata_service = metadata_service

    def search_files(self, query: str, filters: Optional[SearchFilters] = None) -> List[FileModel]:
        """
        Searches for files based on a search query and optional filters

        Args:
            query: The search query
            filters: Optional filters to apply

        Returns:
            List of matching files
        """
        # Initialize filters if not provided
        if filters is None:
            filters = SearchFilters()

        # Get all files that match the basic criteria
        files = self.metadata_service.search_files(
            search_term=query,
            tags=filters.tags,
            mime_types=filters.mime_types
        )

        # Apply additional filtering and ranking
        return self._filter_and_rank_results(files, query, filters)

    def _filter_and_rank_results(
            self, files: List[FileModel], query: str, filters: SearchFilters
    ) -> List[FileModel]:
        """
        Applies additional filtering and ranks search results

        Args:
            files: The initial list of files
            query: The search query
            filters: The search filters

        Returns:
            Filtered and ranked list of files
        """
        # Apply date filters
        filtered_files = []
        for file in files:
            try:
                upload_time = datetime.fromisoformat(file.uploaded_at)
            except (ValueError, TypeError):
                upload_time = datetime.now()

            after_filter = True
            if filters.uploaded_after:
                after_filter = upload_time >= filters.uploaded_after

            before_filter = True
            if filters.uploaded_before:
                before_filter = upload_time <= filters.uploaded_before

            if after_filter and before_filter:
                filtered_files.append(file)

        # Apply size filters
        size_filtered_files = []
        for file in filtered_files:
            min_size_filter = True
            if filters.min_size is not None:
                min_size_filter = file.size >= filters.min_size

            max_size_filter = True
            if filters.max_size is not None:
                max_size_filter = file.size <= filters.max_size

            if min_size_filter and max_size_filter:
                size_filtered_files.append(file)

        # Apply category filters if specified
        if filters.categories:
            category_filtered_files = []
            for file in size_filtered_files:
                if any(
                        category.lower() in [c.lower() for c in file.metadata.categories]
                        for category in filters.categories
                ):
                    category_filtered_files.append(file)
            result_files = category_filtered_files
        else:
            result_files = size_filtered_files

        # Calculate relevance score for each file and sort by it
        scored_files = [
            (file, self._calculate_relevance_score(file, query))
            for file in result_files
        ]
        scored_files.sort(key=lambda x: x[1], reverse=True)

        return [file for file, _ in scored_files]

    def _calculate_relevance_score(self, file: FileModel, query: str) -> float:
        """
        Calculates a relevance score for a file based on how well it matches the query

        Args:
            file: The file model
            query: The search query

        Returns:
            A relevance score (higher is better)
        """
        if not query:
            return 1.0  # No query means all files are equally relevant

        score = 0.0
        query_lower = query.lower()
        query_terms = re.split(r'\s+', query_lower)

        # Check filename match (highest weight)
        file_name_lower = file.name.lower()
        if query_lower in file_name_lower:
            score += 10.0

        # Add points for each query term in the filename
        for term in query_terms:
            if term in file_name_lower:
                score += 5.0

        # Check tags match
        for tag in file.tags:
            tag_lower = tag.lower()
            if query_lower in tag_lower:
                score += 3.0

            for term in query_terms:
                if term in tag_lower:
                    score += 1.5

        # Check categories match
        for category in file.metadata.categories:
            category_lower = category.lower()
            if query_lower in category_lower:
                score += 3.0

            for term in query_terms:
                if term in category_lower:
                    score += 1.5

        # Check extracted text match
        if file.metadata.extracted_text:
            text_lower = file.metadata.extracted_text.lower()

            # Add points for exact matches
            exact_matches = len(re.findall(re.escape(query_lower), text_lower))
            score += exact_matches * 2.0

            # Add points for individual term matches
            for term in query_terms:
                term_matches = len(re.findall(re.escape(term), text_lower))
                score += term_matches * 0.5

        # Check entity matches
        for entity in file.metadata.entities:
            entity_text_lower = entity.text.lower()
            if query_lower in entity_text_lower:
                score += 2.0

            for term in query_terms:
                if term in entity_text_lower:
                    score += 1.0

        return score